src/workerd/api/actor.h

```diff
@@ -13,27 +13,101 @@
 #include <capnp/compat/http-over-capnp.h>
 #include <workerd/api/http.h>
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/function.h>
 
 namespace workerd {
   template <typename T> class IoOwn;
 }
 
 namespace workerd::api {
 
+enum IsInHouse {
+  YES,
+  NO,
+};
+
+// This class is the client RPC stub for Durable Objects. It intercepts dynamic method calls, then
+// sends the method name and arguments to the remote Durable Object, returning the response it
+// receives.
+//
+// We don't want name interception on Fetcher itself -- at least not yet -- since not all Fetcher
+// implementations need it. Name interception is off by default; to turn it on, set the JsRpcStub
+// compatibility flag.
+class DurableObjectRpcStub : public Fetcher, public jsg::NamedIntercept {
+public:
+  DurableObjectRpcStub(
+      IoOwn<OutgoingFactory> outgoingFactory,
+      RequiresHostAndProtocol requiresHost,
+      IsInHouse inHouse,
+      CompatibilityFlags::Reader flags)
+    : Fetcher(kj::mv(outgoingFactory), requiresHost, (inHouse == YES)), flags(flags) {}
+
+  bool isRegisteredName(kj::StringPtr name) {
+    // We don't need to check `fetch` or `constructor` because `fetch` is known at compile time and
+    // `constructor` is a JS keyword; neither uses name interception. This needs to be updated
+    // whenever a new "registered name", aka handler, is added to Durable Objects.
+    if (name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  kj::Maybe<jsg::JsValue> getNamed(jsg::Lock& js, kj::StringPtr name) override {
+    if (flags.getJsRpcStub()) {
+      JSG_REQUIRE(!isRegisteredName(name), Error, "cannot call registered method '", name, "'");
+
+      // Named intercept is enabled, this means we won't default to legacy behavior.
+      // TODO(soon): The server side PR will fill this function with the RPC call and other stuff,
+      //             this is just a temporary placeholder for testing.
+      return jsg::JsValue(js.wrapReturningFunction(js.v8Context(),
+          [](jsg::Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) {
+            return js.str("foo"_kj);
+          }
+      ));
+    }
+    // JS RPC has not been enabled, default to legacy behavior!
+    return kj::none;
+  }
+
+  kj::Array<kj::String> listNamed(jsg::Lock& js) override {
+    // TODO(soon): This will return an empty array for JS RPC because we don't have any dynamic
+    // names beyond the user defined ones, which we obviously can't know in advance.
+    return kj::arr(kj::str("foo"));
```

Comment:
Could we do an actual function here like `jsRpcStubEnabled`? I'm thinking something simple that we wouldn't mind having even when the server side is set up. Also, perhaps return an empty array if `!flags.getJsRpcStub()`?

Reply:
Agreed for empty array condition, but I'm curious about having an actual function returned here. We would need to make a special exception for `jsRpcStubEnabled` in `getNamed()` going forward, and it would be a bit of an awkward exception.

By nature of it being a dynamic name, if the `jsRpcStub` flag is present it will always return true, and if it's absent `stub.jsRpcStubEnabled()` will throw an exception (unless we make it a _really_ special case in `getNamed()`, by ignoring the compat flag if we called `jsRpcStubEnabled`).

Feels a bit weird to me to use the name interception stuff to check for a compat flag that confirms if the stub is allowed to use name interception :sweat_smile:. I'd lean more towards the usual `JSG_METHOD`, but still not sure it's really necessary to know at runtime?

Reply:
Why not just expose it as a non-intercepted property on the same class? Explicitly defined properties (e.g. using `JSG_READONLY_PROTOTYPE_PROPERTY`, etc) should still Just Work here also.

Reply:
Ah, sounds like maybe it's not worth it then. I was just thinking of how to exercise `listNamed()` reasonably, so maybe the answer is to return an empty array now if we can?